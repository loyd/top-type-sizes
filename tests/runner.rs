use top_type_sizes::*;

macro_rules! test_sample {
    ($name:ident) => {
        #[test]
        fn $name() {
            static SAMPLE: &'static str =
                include_str!(concat!("samples/", stringify!($name), ".txt"));

            let refined_sample = reader::read(SAMPLE.as_bytes()).unwrap();
            let types = parser::parse(&refined_sample).unwrap();
            insta::assert_debug_snapshot!(format!("{}_types", stringify!($name)), types);

            // TODO: check the transformer.

            let output = formatter::format(types, &Default::default());
            println!("{output}");
            // TODO: make snapshots of output.
        }
    };
}

test_sample!(command);
test_sample!(control_flow);
test_sample!(alignment_enum);
test_sample!(tokio_udp);
test_sample!(several_types);
test_sample!(timex);
test_sample!(async_fn);
test_sample!(compiler_messages);

// TODO: add samples from rustc tests.
