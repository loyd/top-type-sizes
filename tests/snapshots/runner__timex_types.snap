---
source: tests/runner.rs
expression: types
---
[
    Type {
        name: "unix::linux_like::linux::gnu::timex",
        size: 208,
        align: 8,
        kind: Struct(
            StructType {
                items: [
                    Field(
                        Field {
                            kind: AdtField,
                            name: "modes",
                            size: 4,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Padding(
                        4,
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "offset",
                            size: 8,
                            align: Some(
                                8,
                            ),
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "freq",
                            size: 8,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "maxerror",
                            size: 8,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "esterror",
                            size: 8,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "status",
                            size: 4,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Padding(
                        4,
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "constant",
                            size: 8,
                            align: Some(
                                8,
                            ),
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "precision",
                            size: 8,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "tolerance",
                            size: 8,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "time",
                            size: 16,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "tick",
                            size: 8,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "ppsfreq",
                            size: 8,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "jitter",
                            size: 8,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "shift",
                            size: 4,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Padding(
                        4,
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "stabil",
                            size: 8,
                            align: Some(
                                8,
                            ),
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "jitcnt",
                            size: 8,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "calcnt",
                            size: 8,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "errcnt",
                            size: 8,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "stbcnt",
                            size: 8,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "tai",
                            size: 4,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "__unused1",
                            size: 4,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "__unused2",
                            size: 4,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "__unused3",
                            size: 4,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "__unused4",
                            size: 4,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "__unused5",
                            size: 4,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "__unused6",
                            size: 4,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "__unused7",
                            size: 4,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "__unused8",
                            size: 4,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "__unused9",
                            size: 4,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "__unused10",
                            size: 4,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                    Field(
                        Field {
                            kind: AdtField,
                            name: "__unused11",
                            size: 4,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                ],
            },
        ),
        end_padding: None,
    },
]
